torch/_dynamo/side_effects.py

```diff
@@ -375,6 +380,54 @@ def codegen_save_tempvars(self, cg: PyCodegen):
                 ]
             )
 
+    def register_hook(self, tensor, hook, handle):
+        if tensor not in self.tensor_hooks:
+            self.tensor_hooks[tensor] = ([], False)
+        self.tensor_hooks[tensor][0].append((hook, handle))
+
+    def codegen_hooks(self, cg, tensor):
+        if tensor not in self.tensor_hooks:
+            return
+        for (
+            hook,
+            handle,
+        ) in self.tensor_hooks[
+            tensor
+        ][0]:
+            # On dynamo tensor_hooks
+            #
+            # register_hook in the Graph: We bypass direct inclusion of register_hook calls in the graph.
+            # Instead, these are tracked and stashed as a global variable, enabling their association with tensors in
+            # the residuals. During dynamo's frame creation, these hooks are invoked seamlessly.
+            #
+            # Handling the Handle: When a user retains the register_hook result in a handle, we intercept the
+            # STORE_FAST operation to record the user-designated local variable name. This ensures the reconstructed
+            # bytecode retains this name. If no handle is defined, we simply pop the generated value to keep the
+            # stack intact.
+            #
+            # Dynamo Tensor Hooks Workflow:
+            # - Functions passed to register_hook are lifted globally.
+            # - In the "side_effects" phase of codegen, we iterate over tensors with hooks to:
+            #   - Generate the tensor.
+            #   - Issue a register_hook call on the tensor, linking to the globally stored function.
+            #   - Incorporate a handle if one was established in the eager phase.
+            # The handle's exact user-specified name, "last_seen_name", is discerned and associated during STORE_FAST.
+            # return
+            cg(tensor)
```

Comment:
Self note - may need a different load() mechanism, can cause re-entrancy bugs here. Should be easy to pass in a load fn for output cases where tensor is already loaded.